esmvalcore/_recipe.py

```diff
@@ -475,13 +492,27 @@ def _update_fx_settings(settings, variable, config_user):
         settings['weighting_landsea_fraction']['fx_files'] = fx_dict
         logger.info(msg, 'land/sea fraction weighting', pformat(fx_dict))
 
-    for step in ('area_statistics', 'volume_statistics'):
+    for step in ('area_statistics', 'volume_statistics', 'zonal_statistics'):
+        var = dict(variable)
+        fx_files_dict = {}
         if settings.get(step, {}).get('fx_files'):
-            var = dict(variable)
             var['fx_files'] = settings.get(step, {}).get('fx_files')
-            fx_files_dict = {
-                fxvar: _get_correct_fx_file(variable, fxvar, config_user)
-                for fxvar in var['fx_files']}
+            for fxvar in var['fx_files']:
+                fxvar_name = fxvar
+                if isinstance(fxvar, dict):
+                    fxvar_name = fxvar.get("short_name")
+                    if not fxvar_name:
+                        raise RecipeError(
+                            "Key short_name missing from {}".format(fxvar))
+                _, cmor_fx_var = _get_correct_fx_file(
+                    variable, fxvar, config_user)
+                if cmor_fx_var:
+                    fx_files_dict[fxvar_name] = get_output_file(
+                        var,
+                        config_user['preproc_dir'],
+                        fx_var_alias=cmor_fx_var)
+
+            # finally construct the fx files dictionary
             settings[step]['fx_files'] = fx_files_dict
             logger.info(msg, step, pformat(fx_files_dict))
```

Comment:
there is no actual duplication, if you look carefully, there are a lot of differences between the cases both at logger and functional levels. However, I broke down the big function in two parts now, that deal with totally separate things

Reply:
There absolutely **is** duplication: line 474 is the same as 491, line 476 is almost the same as 492 and it smells strongly like they should be identical. Indeed, the structure I proposed above would go some way towards that since the `step_settings` would deal with the differences in the lvalue. That leaves the difference that one assigns a dict, the other a list of dicts. Furthermore, the code of `_get_landsea_fraction_fx_dict` looks suspiciously similar to the code block in the `'mask_landseaice'` branch. Probably all three cases can be dealt with in one function similar to an extended version of `_get_landsea_fraction_fx_dict` that takes a list of fx variables as an additional argument.

Reply:
gonna try change now :beer:

Reply:
what do you think about it now @zklaus -man?

Reply:
I put a more fleshed out suggestion in #522. Have a look and tell me what you think. :beers:

Reply:
Shall do in a wee bit, need coffee now. Cheers, man :beer:

Reply:
#522 merged in here so am resolving the convo :beer:

Reply:
Ok. I'm putting the open conversation in the comments here then.